src/imputors.jl

```diff
@@ -8,39 +8,75 @@ implement the `impute!(imp::<MyImputor>, data::AbstractVector)` method.
 abstract type Imputor end
 
 # A couple utility methods to avoid messing up var and obs dimensions
-obsdim(imp::Imputor) = imp.vardim == 1 ? 2 : 1
-vardim(imp::Imputor) = imp.vardim
+obsdim(dims) = dims
+vardim(dims) = dims == 1 ? 2 : 1
 
-function obswise(imp::Imputor, data::AbstractMatrix)
-    return (selectdim(data, obsdim(imp), i) for i in axes(data, obsdim(imp)))
+function obswise(data::AbstractMatrix; dims=1)
+    return (selectdim(data, obsdim(dims), i) for i in axes(data, obsdim(dims)))
 end
 
-function varwise(imp::Imputor, data::AbstractMatrix)
-    return (selectdim(data, vardim(imp), i) for i in axes(data, vardim(imp)))
+function varwise(data::AbstractMatrix; dims=2)
+    return (selectdim(data, vardim(dims), i) for i in axes(data, vardim(dims)))
 end
 
-function filterobs(f::Function, imp::Imputor, data::AbstractMatrix)
-    mask = [f(x) for x in obswise(imp, data)]
-    return imp.vardim == 1 ? data[:, mask] : data[mask, :]
+function filterobs(f::Function, data::AbstractMatrix; dims=1)
+    mask = [f(x) for x in obswise(data; dims=dims)]
+    return dims == 1 ? data[mask, :] : data[:, mask]
 end
 
-function filtervars(f::Function, imp::Imputor, data::AbstractMatrix)
-    mask = [f(x) for x in varwise(imp, data)]
-    return imp.vardim == 1 ? data[mask, :] : data[:, mask]
+function filtervars(f::Function, data::AbstractMatrix; dims=2)
+    mask = [f(x) for x in varwise(data; dims=dims)]
+    return dims == 1 ? data[:, mask] : data[mask, :]
 end
 
 """
-    impute(data, imp::Imputor)
+    splitkwargs(::Type{T}, kwargs...) where T <: Imputor -> (imp, rem)
+
+Takes an Imputor type with kwargs and returns the constructed imputor and the
+unused kwargs which should be passed to the `impute!` call.
+
+NOTE: This is used by utility methods with construct and imputor and call impute in 1 call.
+"""
+function splitkwargs(::Type{T}, kwargs...) where T <: Imputor
+    rem = Dict(kwargs...)
+    kwdef = empty(rem)
+
+    for f in fieldnames(T)
+        if haskey(rem, f)
+            kwdef[f] = rem[f]
+            delete!(rem, f)
+        end
+    end
+
+    return (T(; kwdef...), rem)
+end
+
+# Some utility methods for constructing imputors and imputing data in 1 call.
+function impute(data, t::Type{T}, kwargs...) where T <: Imputor
+    imp, rem = splitkwargs(t, _extract_context_kwargs(kwargs...)...)
+    return impute(data, imp; rem...)
+end
+
+function impute!(data, t::Type{T}, kwargs...) where T <: Imputor
+    imp, rem = splitkwargs(t, _extract_context_kwargs(kwargs...)...)
+    return impute!(data, imp; rem...)
+end
```

Comment:
Why would we want to do this? It seems weird to me to call `impute` with the type rather than an instance of the type. 

Why would we e.g. call `impute(X, DropObs, context=context, dims=dims)` rather than `impute(X, DropObs(context=context), dims=dims)`? That seems to muddy things, to me, so perhaps you have another use-case in mind?

we already have `Impute.<type>` functions like `Impute.fill` for easier use, and I'm wary of too many ways to do the same thing

Reply:
This is only intended for internal use. We could add an underscore to make that clear if you like.

Reply:
ah, cool -- yes, that'd be good -- I was confused by the API for a sec there :)